src/tyr/route_serializer_osrm.cc

```diff
@@ -1454,6 +1456,99 @@ void maneuver_geometry(json::MapPtr& step,
   }
 }
 
+// The idea is that the instructions come a fixed amount of seconds before the maneuver takes place.
+// For whatever reasons, a distance in meters from the end of the maneuver needs to be provided
+// though. When different speeds are used on the road, they all need to be taken into account. This
+// function calculates the distance before the end of the maneuver by checking the elapsed_cost
+// seconds of each edges and accumulates their distances until the seconds threshold is passed. The
+// speed of this last edge is then used to subtract the distance so that the the seconds until the end
+// are exactly the provided amount of seconds.
+float distance_along_geometry(const valhalla::DirectionsLeg::Maneuver* prev_maneuver,
+                              valhalla::odin::EnhancedTripLeg* etp,
+                              const double distance,
+                              const uint32_t target_seconds) {
+  uint32_t node_index = prev_maneuver->end_path_index();
+  double end_node_elapsed_seconds = etp->node(node_index).cost().elapsed_cost().seconds();
+  double begin_node_elapsed_seconds =
+      etp->node(prev_maneuver->begin_path_index()).cost().elapsed_cost().seconds();
+
+  // If the maneuver is too short, simply return its distance.
+  if (end_node_elapsed_seconds - begin_node_elapsed_seconds < target_seconds) {
+    return distance;
+  }
+
+  float accumulated_distance_km = 0;
+  double accumulated_seconds = 0;
+  // Find the node after which the instructions should be heard:
+  while (accumulated_seconds < target_seconds && node_index >= prev_maneuver->begin_path_index()) {
+    node_index -= 1;
+    // not really accumulating seconds ourselves, but it happens elsewhere:
+    accumulated_seconds =
+        end_node_elapsed_seconds - etp->node(node_index).cost().elapsed_cost().seconds();
+    accumulated_distance_km += etp->GetCurrEdge(node_index)->length_km();
+  }
+  // The node_index now indicates the node AFTER which the target_seconds will be reached
+  // we now have to subtract the surplus seconds * speed of this edge from the accumulated_distance_km
+  float speed = etp->GetCurrEdge(node_index)->speed(); // speed is in km/h
+  float hours_to_be_removed = (accumulated_seconds - target_seconds) / 3600;
+  accumulated_distance_km -= speed * hours_to_be_removed;
```

Comment:
this worries me a little bit because there are a lot of factors that comprise this final value:

https://github.com/valhalla/valhalla/blob/5540cd7d0fcf67b5a1b8ee16fc5d6cbd2360b4c9/src/thor/triplegbuilder.cc#L1135-L1152

i could imagine or i am a tiny bit worried that we'll see a voice instruction at some crazy spot along the maneuver and find a gotcha here somewhere. cant we just do a linear combination of the edge we stop on?

```suggestion
  float accumulated_distance_km = 0;
  float previous_accumulated_distance_km = 0;
  double accumulated_seconds = 0;
  double previous_accumulated_seconds = 0;
  // Find the node after which the instructions should be heard:
  while (accumulated_seconds < target_seconds && node_index >= prev_maneuver->begin_path_index()) {
    node_index -= 1;
    // not really accumulating seconds ourselves, but it happens elsewhere:
    previous_accumulated_seconds = accumulated_seconds;
    accumulated_seconds =
        end_node_elapsed_seconds - etp->node(node_index).cost().elapsed_cost().seconds();
    previous_accumulated_distance_km = accumulated_distance_km;
    accumulated_distance_km += etp->GetCurrEdge(node_index)->length_km();
  }
  // The node_index now indicates the node AFTER which the target_seconds will be reached
  // we now have to subtract the surplus distance (based on seconds) of this edge from the accumulated_distance_km
  auto surplus_percentage = (accumulated_seconds - target_seconds) / (accumulated_seconds - previous_accumulated_seconds);
  accumulated_distance_km -= (accumulated_distance_km - previous_accumulated_distance_km) * surplus_percentage;
```

Reply:
My (granted limited) understanding of the elapsed time is that it is the quotient of the edge length divided by the edge speed. So getting the remaining distance along the edge would simply be the product of the speed times the remaining time.

What you are linking to suggests, that it is the other way around, the speed is calculated by dividing the length by the elapsed time (if at all). And your worry is that, the speed is not the best value to rely on. Instead you would suggest, that remaining distance along the edge should be calculated based on the length and elapsed time like this:

```
edge_duration = elapsed_time_at_the_end - elapsed_time_at_the_beginning
distance_along_edge = edge_length * remaining_time / edge_duration
```

Correct?

Reply:
My (granted limited) understanding of the elapsed time is that it is the quotient of the edge length divided by the edge speed. So getting the remaining distance along the edge would simply be the product of the speed times the remaining time.

What you are linking to suggests, that it is the other way around, the speed is calculated by dividing the length by the elapsed time (if at all). And your worry is that, the speed is not the best value to rely on. Instead you would suggest, that remaining distance along the edge should be calculated based on the length and elapsed time like this:

```
edge_duration = elapsed_time_at_the_end - elapsed_time_at_the_beginning
distance_along_edge = edge_length * remaining_time / edge_duration
```

Correct?

Reply:
I applied your suggested changes and made sure the tests still pass and applied clang-format to it.